README.md
For an example, see the [Michigan Genomics Initiative PheWeb](http://pheweb.sph.umich.edu).
For a walk-through demo see [here](etc/demo.md#demo-navigating-pheweb).
If you have questions or comments, check out our [Google Group](https://groups.google.com/g/pheweb-umich).

![screenshot of PheWAS plot](https://cloud.githubusercontent.com/assets/862089/25474725/3edbe256-2b02-11e7-8abb-0ca26d406b11.png)

# How to Cite PheWeb
If you use the PheWeb code base for your work, please cite our paper:

Gagliano Taliun, S.A., VandeHaar, P. et al. Exploring and visualizing large-scale genetic associations by using PheWeb. *Nat Genet* 52, 550–552 (2020).

# How to Build a PheWeb for your Data

If this is broken, [open an issue on github](https://github.com/statgen/pheweb/issues/new) and hopefully I can help.

### 1. Install PheWeb

```bash
pip3 install pheweb
```

- If that doesn't work, follow [the detailed install instructions](etc/detailed-install-instructions.md#detailed-install-instructions).

### 2. Create a directory and `config.py` for your new dataset

```
mkdir ~/my-new-pheweb && cd ~/my-new-pheweb
```

This directory will store all the files pheweb makes for your dataset. All `pheweb ...` commands should be run in this directory.

Make `config.py` in this directory. In it, either set `hg_build_number = 19` or `hg_build_number = 38`.  Other options you can set are listed [here](etc/detailed-loading-instructions.md#configuration-options).

### 3. Prepare your association files

You need one file for each phenotype.  Most common GWAS file formats should work.  Here are the requirements:

- It needs a header row.
- Columns can be delimited by tabs, spaces, or commas.
- It needs a column for the reference allele (which must always match the bases on the reference genome that you specified with `hg_build_number`) and a column for the alternate allele.  If you have a `MARKER_ID` column like `1:234_C/G`, that's okay too.  If you have an allele1 and allele2, and sometimes one or the other is the reference, then you'll need to modify your files.
- It can be gzipped if you want.
- Variants must be sorted by chromosome and position, with chromosomes in the order [1-22,X,Y,MT].

The file must have columns for:

| column description | name    | other allowed column names | allowed values |
| ---                | ---     | ---                        | --- |
| chromosome         | `chrom` | `#chrom`, `chr`            | 1-22, `X`, `Y`, `M`, `MT`, `chr1`, etc |
| position           | `pos`   | `beg`, `begin`, `bp`       | integer |
| reference allele   | `ref`   | `reference`                | must match reference genome |
| alternate allele   | `alt`   | `alternate`                | anything |
| p-value            | `pval`  | `pvalue`, `p`, `p.value`   | number in [0,1] |


You may also have columns for:

| column description            | name           | other allowed column names | allowed values |
| ---                           | ---            | ---                        | --- |
| minor allele frequency        | `maf`          |                            | number in (0,0.5] |
| allele frequency              | `af`           | `a1freq`, `frq`            | number in (0,1) |
| AF among cases                | `case_af`      | `af.cases`                 | number in (0,1) |
| AF among controls             | `control_af`   | `af.controls`              | number in (0,1) |
| allele count                  | `ac`           |                            | integer |
| effect size                   | `beta`         |                            | number |
| standard error of effect size | `sebeta`       | `se`                       | number |
| odds ratio                    | `or`           |                            | number |
| R2                            | `r2`           |                            | number |
| number of samples             | `num_samples`  | `ns`, `n`                  | integer, must be the same for every variant in its phenotype |
| number of controls            | `num_controls` | `ns.ctrl`, `n_controls`    | integer, must be the same for every variant in its phenotype |
| number of cases               | `num_cases`    | `ns.case`, `n_cases`       | integer, must be the same for every variant in its phenotype |


Column names are case-insensitive.  If your file has a different column name, set `field_aliases = {"column_name": "field_name"}` in `config.py`.  For example, `field_aliases = {'P_BOLT_LMM_INF': 'pval', 'NSAMPLES': 'num_samples'}`.

Any field can be null if it is one of ['', '.', 'NA', 'N/A', 'n/a', 'nan', '-nan', 'NaN', '-NaN', 'null', 'NULL'].  If a required field is null, the variant gets dropped.


### 4. Make a list of your phenotypes

Inside of your data directory, you need a file named `pheno-list.json` that looks like this:

```json
[
 {
  "assoc_files": ["/home/peter/data/ear-length.gz"],
  "phenocode": "ear-length"
 },
 {
  "assoc_files": ["/home/peter/data/a1c.X.gz","/home/peter/data/a1c.autosomal.gz"],
  "phenocode": "A1C"
 }
]
```

Each phenotype needs `assoc_files` (a list of paths to association files) and `phenocode` (a string representing your phenotype that is used in filenames and URLs, comprised of `[A-Za-z0-9_~-]`).

If you want, you can also include:

- `phenostring` (string): a name for the phenotype. Shown in tables and tooltips and page headers.
- `category` (string): groups together phenotypes in the PheWAS plot. Shown in tables and tooltips.
- `num_cases`, `num_controls`, and/or `num_samples` (number): if your input data only has `AC` or `MAC`, this will be used to calculated `AF` or `MAF`.  Shown in tooltips.  If your input data has correctly-named columns for these, the command `pheweb phenolist read-info-from-association-files` will add them into your existing `pheno-list.json`.
- anything else you want, but you'll have to modify templates to use it.

You can use a csv by running:

```
pheweb phenolist import-phenolist "/path/to/pheno-list.csv"
```

or you can make one from scratch by running:

```
pheweb phenolist glob-files --star-is-phenocode "/home/peter/data/*.gz"
```

You can see other methods [here](etc/detailed-loading-instructions.md#making-pheno-listjson).


### 5. Load your association files

Run `pheweb process`.

To distribute jobs across a cluster, follow [these instructions](etc/detailed-loading-instructions.md#distributing-jobs-across-a-cluster).

To include VEP annotations, follow [these instructions](etc/detailed-loading-instructions.md#annotating-with-vep).

If something breaks and you can't understand the error message or it's something that PheWeb should support by default, [open an issue on github](https://github.com/statgen/pheweb/issues/new) or email me.

### 6. Serve the website

Run `pheweb serve --open`.

That command should either open a browser to your new PheWeb, or it should give you a URL that you can open in your browser to access your new PheWeb.
If it doesn't, follow [the directions for hosting a PheWeb and accessing it from your browser](etc/detailed-webserver-instructions.md#hosting-a-pheweb-and-accessing-it-from-your-browser).

### More options:

To run pheweb through systemd, see sample file [here](etc/pheweb.service).
To use Apache2 or Nginx, see instructions [here](etc/detailed-webserver-instructions.md#using-apache2-or-nginx).
To require login via OAuth, see instructions [here](etc/detailed-webserver-instructions.md#using-oauth).
To track page views with Google Analytics, see instructions [here](etc/detailed-webserver-instructions.md#using-google-analytics).
To reduce storage use, see instructions [here](etc/detailed-webserver-instructions.md#reducing-storage-use).
To customize page contents, see instructions [here](etc/detailed-webserver-instructions.md#customizing-page-contents).

PheWeb can display phenotype correlations generated by [another tool](https://github.com/statgen/pheweb-rg-pipeline).
To use this feature, set `show_correlations = True`  in `config.py` and place the output of the rg pipeline as `pheno-correlations.txt` in the same folder as `pheno-list.json`.

To hide the button for downloading summary stats, add `download_pheno_sumstats = "secret"` and `SECRET_KEY = "your random string"` in `config.py`.  That will make a secret page (printed to the console when you start the server) to share summary stats.

To allow dynamically filtering the manhattan plot, run `pheweb best-of-pheno` and set `show_manhattan_filter_button=True` in `config.py`.

# Modifying PheWeb

See instructions [here](etc/detailed-development-instructions.md).
See documentation about the files in `generated-by-pheweb/` [here](etc/detailed-internal-dataflow.md).


CHANGELOG.md
*This file only includes changes that are relevant to people running a pheweb site.*

## 1.3.15
- Fixes `pheweb cluster`.

## 1.3.14
- Fixes uppercase `field_aliases` in `config.py`.  Column names are case-insensitive now.

## 1.3.13
- Speeds up autocomplete

## 1.3.12
- Adds beta/sebeta columns to the tables on /pheno/ and /variant/
- Shows AF range or MAF range better on /variant/
- Shows pvalue=0 as p<1e-320 in most places.
- Improves error-handling on /pheno-filter/
- Upgrades to LocusZoom.js 0.13, including new PNG downloads
- Fixes bugs in OAuth and WSGI
- Uses relative redirects, so that http vs https and hostname don't matter, except in OAuth code.

## 1.3.9
- Improves hovering on the filtered manhattan plots
- Includes code for annotating with VEP
- Shows category on /top_hits

**Changes needed to data:**

- Run `rm generated-by-pheweb/top_hits.json; pheweb top-hits`

## 1.3.7
- Uses gencode v37 (released 2021-Feb)
- Shows GClambda and num_samples/num_cases/num_controls and num_loci<5e8 on /phenotypes
- Supports custom_templates/ again

**Changes needed to data:**

- Run `rm generated-by-pheweb/sites/sites.tsv && pheweb process` (because gene names must agree beween autocompletion and the pre-processed data)

## 1.3.6
- Speeds up `pheweb gather-pvalues-for-each-gene` ~2x by avoiding reading any variant twice.  (Thanks to finngen for this suggestion.)
- Allows live-filtering a manhattan plot by MAF or snp/indel, with instructions in README.
- Verifies that `num_cases + num_controls == num_samples` in `pheweb phenolist verify` (which is included in `pheweb process`).

## 1.3.5
- Removes dependence on `pandas` (because it wouldn't install on my laptop)

## 1.3.4
- Allows setting `loading_nice = True`.
- Allows setting `field_aliases` again.
- Reduces memory usage by `pheweb qq` by ~10x by switching to `numpy` and `pandas`.
- Fixes the bug where `pheweb matrix` breaks when `matrix.tsv.gz` is up-to-date.

## 1.3.0
- Rewrites configuration management, losing the ability to customize `extra_per_*_fields` and `null_values` and `field_aliases`.
- Fixes bug where config wasn't passed to child processes when using `PHEWEB_DATADIR` or `pheweb conf key=value <subcommand>`.

Bugs:

- `pheweb matrix` breaks when `matrix.tsv.gz` is already up-to-date.

## 1.2.5
- Makes sure that `pheno_gz/<phenocode>.gz.tbi` gets created, and re-runs traits that don't have it.

## 1.2.3
- Uses dbSNP v154 (the latest!) with way more rsids.  To use them, run `rm generated-by-pheweb/sites/sites-rsids.tsv && pheweb process`.

## 1.2.1
- Allows hg38 via `hg_build_number=38`
- Downloads resources from <https://resources.pheweb.org> instead of processing raw data from EBI, dbSNP, etc.
- Replaces marisa-trie with sqlite3 to remove a flaky dependency and improve the order of autocomplete suggestions.
- Replaces more json files with sqlite3 to handle large datasets better.
- Compresses all internal files with `gzip -2` to save storage and IO.
- Gets rid of `generated-by-pheweb/pheno/`, relying on `generated-by-pheweb/pheno_gz/` instead.
- Allows `chr1`-`chr25` in input files.

**Changes needed to data:**

- Run `pheweb download-genes`
- Run `pheweb make-gene-aliases-sqlite3`
- Run `rm generated-by-pheweb/phenotypes.json; pheweb phenotypes`
- Run `pheweb gather-pvalues-for-each-gene`

## 1.2.0 (broken)
Bugs:

- `pheweb matrix` fails to match filenames to columns.

## 1.1.28
- Allows selecting which phenotypes to run in most steps via `pheweb <subcommand> --phenos=5-10`.
- Adds `pheweb cluster --step=<subcommand>`.


utils.py

from __future__ import print_function, division, absolute_import

# Load config
import os
import imp
possible_conf_files = [
    os.path.join(os.path.abspath(os.curdir), 'config.py'),
    os.path.join(os.path.dirname(os.path.abspath(__file__)), 'config.py'),
]
if 'PHEWEB_DATADIR' in os.environ:
    possible_conf_files.insert(0, os.path.join(os.environ['PHEWEB_DATADIR'], 'config.py'))
conf = None
for conf_file in possible_conf_files:
    if os.path.exists(conf_file):
        conf = imp.load_source('conf', conf_file)
        break
if conf is None:
    raise Exception('PheWeb failed to find a config.py file.  Places checked:\n{}'.format('\n'.join(possible_conf_files)))

# Prepare config
if not hasattr(conf, 'data_dir'):
    conf.data_dir = os.path.dirname(conf.__file__)
if not os.access(conf.data_dir, os.R_OK):
    raise Exception("It looks like you don't have permission to read your data_dir, {!r}".format(conf.data_dir))
if not hasattr(conf, 'source_file_parser'): # TODO: rename `association_file_parser`, relegate source_file_parser to an alias
    conf.source_file_parser = 'epacts'
if hasattr(conf, 'cache_dir'):
    if not conf.cache_dir:
        del conf.cache_dir
    else:
        conf.cache_dir = os.path.expanduser(conf.cache_dir)
        if not os.access(conf.cache_dir, os.R_OK):
            print('Warning: the directory {!r} is configured to be your cache directory in {!r} but it is not readable'.format(
                conf.cache_dir, conf_file))


import re
import itertools
import functools
import traceback
import math
import json
import gzip
import os
import errno
import random
import sys
import subprocess
import time


def parse_variant(query, default_chrom_pos = True):
    if isinstance(query, unicode):
        query = query.encode('utf-8')
    chrom_pattern = r'(?:[cC][hH][rR])?([0-9XYMT]+)'
    chrom_pos_pattern = chrom_pattern + r'[-_:/ ]([0-9]+)'
    chrom_pos_ref_alt_pattern = chrom_pos_pattern + r'[-_:/ ]([-AaTtCcGg\.]+)[-_:/ ]([-AaTtCcGg\.]+)'

    match = re.match(chrom_pos_ref_alt_pattern, query) or re.match(chrom_pos_pattern, query) or re.match(chrom_pattern, query)
    g = match.groups() if match else ()

    if default_chrom_pos:
        if len(g) == 0: g += ('1',)
        if len(g) == 1: g += (0,)
    if len(g) >= 2: g = (g[0], int(g[1])) + tuple([bases.upper() for bases in g[2:]])
    return g + tuple(itertools.repeat(None, 4-len(g)))



def parse_marker_id(marker_id):
    match = parse_marker_id.regex.match(marker_id)
    if match is None:
        raise Exception("ERROR: MARKER_ID didn't match our MARKER_ID pattern: {!r}".format(marker_id))
    chrom, pos, ref, alt = match.groups()
    return chrom, int(pos), ref, alt
parse_marker_id.regex = re.compile(r'([^:]+):([0-9]+)_([-ATCG\.]+)/([-ATCG\.]+)')


def round_sig(x, digits):
    return 0 if x==0 else round(x, digits-1-int(math.floor(math.log10(abs(x)))))
assert round_sig(0.00123, 2) == 0.0012
assert round_sig(1.59e-10, 2) == 1.6e-10

def get_phenolist():
    fname = os.path.join(conf.data_dir, 'pheno-list.json')
    try:
        with open(os.path.join(fname)) as f:
            return json.load(f)
    except IOError: # TODO: these exceptions change in python3
        die("You need a file to define your phenotypes at '{fname}'.\n".format(fname=fname) +
                  "For more information on how to make one, see <https://github.com/statgen/pheweb#3-make-a-list-of-your-phenotypes>")
    except ValueError:
            print("Your file at '{fname}' contains invalid json.\n".format(fname=fname) +
                  "The error it produced was:")
            raise

def get_phenos_with_colnums(app_root_path):
    phenos_by_phenocode = {pheno['phenocode']: pheno for pheno in get_phenolist()}
    with gzip.open(conf.data_dir + '/matrix.tsv.gz') as f:
        header = f.readline().rstrip('\r\n').split('\t')
    assert header[:7] == '#chrom pos ref alt rsids nearest_genes maf'.split()
    for phenocode in phenos_by_phenocode:
        phenos_by_phenocode[phenocode]['colnum'] = {}
    for colnum, colname in enumerate(header[7:], start=7):
        label, phenocode = colname.split('@')
        phenos_by_phenocode[phenocode]['colnum'][label] = colnum
    for phenocode in phenos_by_phenocode:
        assert 'pval' in phenos_by_phenocode[phenocode]['colnum'], (phenocode, phenos_by_phenocode[phenocode])
    return phenos_by_phenocode


pheno_fields_to_include_with_variant = {
    'phenostring', 'category', 'num_cases', 'num_controls', 'num_samples',
}

def get_variant(query, phenos):
    import pysam
    # todo: differentiate between parse errors and variants-not-found
    chrom, pos, ref, alt = parse_variant(query)
    assert None not in [chrom, pos, ref, alt]

    tabix_file = pysam.TabixFile(conf.data_dir + '/matrix.tsv.gz')
    tabix_iter = tabix_file.fetch(chrom, pos-1, pos+1, parser = pysam.asTuple())
    for variant_row in tabix_iter:
        if int(variant_row[1]) == int(pos) and variant_row[3] == alt:
            matching_variant_row = tuple(variant_row)
            break
    else: # didn't break
        return None

    maf = round_sig(float(matching_variant_row[6]), 3)
    assert 0 < maf <= 0.5

    rv = {
        'variant_name': '{} : {:,} {}>{}'.format(chrom, pos, ref, alt),
        'chrom': chrom,
        'pos': pos,
        'ref': ref,
        'alt': alt,
        'maf': maf,
        'rsids': matching_variant_row[4],
        'nearest_genes': matching_variant_row[5],
        'phenos': [],
    }

    for phenocode, pheno in phenos.iteritems():
        try:
            pval = float(matching_variant_row[pheno['colnum']['pval']])
        except ValueError:
            pval = 1
        rv['phenos'].append({
            'phenocode': phenocode,
            'pval': pval,
        })
        for key in pheno:
            if key in pheno_fields_to_include_with_variant:
                rv['phenos'][-1][key] = pheno[key]
    return rv


def mkdir_p(path):
    # like `mkdir -p`
    try:
        os.makedirs(path)
    except OSError as exc:
        if exc.errno != errno.EEXIST or not os.path.isdir(path):
            raise

def activate_virtualenv():
    if hasattr(conf, 'virtualenv_dir'):
        path = os.path.join(conf.virtualenv_dir, 'bin/activate_this.py')
        with open(path) as f:
            code = compile(f.read(), path, 'exec')
        exec(code, dict(__file__=path))

def get_random_page():
    with open(os.path.join(conf.data_dir, 'top_hits.json')) as f:
        hits = json.load(f)
        hits_to_choose_from = [hit for hit in hits if hit['pval'] < 5e-8]
        if not hits_to_choose_from:
            hits_to_choose_from = hits
        if not hits:
            return None
    hit = random.choice(hits_to_choose_from)
    r = random.random()
    if r < 0.4:
        return '/pheno/{}'.format(hit['phenocode'])
    elif r < 0.8:
        return '/variant/{chrom}-{pos}-{ref}-{alt}'.format(**hit)
    else:
        offset = int(50e3)
        return '/region/{phenocode}/{chrom}:{pos1}-{pos2}'.format(pos1=hit['pos']-offset, pos2=hit['pos']+offset, **hit)

def die(message):
    print(message, file=sys.stderr)
    raise Exception()

def exception_printer(f):
    @functools.wraps(f)
    def f2(*args, **kwargs):
        try:
            rv = f(*args, **kwargs)
        except Exception as exc:
            time.sleep(2*random.random()) # hopefully avoid interleaved printing (when using multiprocessing)
            traceback.print_exc()
            print(exc)
            if args: print('args were: {!r}'.format(args))
            if kwargs: print('kwargs were: {!r}'.format(args))
            raise
    return f2

def all_equal(iterator):
    try:
        first = next(iterator)
    except StopIteration:
        return True
    return all(it == first for it in iterator)

def sorted_groupby(iterator, key=None):
    if key is None: key = (lambda v:v)
    return [list(group) for _, group in itertools.groupby(sorted(iterator, key=key), key=key)]

class open_maybe_gzip(object):
    def __init__(self, fname, *args):
        self.fname = fname
        self.args = args
    def __enter__(self):
        is_gzip = False
        with open(self.fname, 'rb') as f:
            if f.read(3) == b'\x1f\x8b\x08':
                is_gzip = True
        if is_gzip:
            self.f = gzip.open(self.fname, *self.args)
        else:
            self.f = open(self.fname, *self.args)
        return self.f
    def __exit__(self, *exc):
        self.f.close()

def pairwise(iterable):
    "s -> (s0, s1), (s2, s3), (s4, s5), ..."
    it = iter(iterable)
    return itertools.izip(it, it)

# TODO: chrom_order_list[25-1] = 'M', chrom_order['M'] = 25-1, chrom_order['MT'] = 25-1 ?
#       and epacts.py should convert all chroms to chrom_idx?
chrom_order_list = [str(i) for i in range(1,22+1)] + ['X', 'Y', 'M', 'MT']
chrom_order = {chrom: index for index,chrom in enumerate(chrom_order_list)}


def get_path(cmd, attr=None):
    if attr is None: attr = '{}_path'.format(cmd)
    path = None
    if hasattr(conf, attr):
        path = getattr(conf, attr)
    else:
        try:
            path = subprocess.check_output(['which', cmd]).strip()
        except subprocess.CalledProcessError:
            pass
    if path is None:
        raise Exception("The command '{cmd}' was not found in $PATH and was not specified (as {attr}) in config.py.".format(cmd=cmd, attr=attr))
    return path

def run_script(script):
    script = 'set -euo pipefail\n' + script
    try:
        with open(os.devnull) as devnull:
            # is this the right way to block stdin?
            data = subprocess.check_output(['sh', '-c', script], stderr=subprocess.STDOUT, stdin=devnull)
        status = 0
    except subprocess.CalledProcessError as ex:
        data = ex.output
        status = ex.returncode
    data = data.decode('utf8')
    if status != 0:
        print('FAILED with status {}'.format(status))
        print('output was:')
        print(data)
        sys.exit(1)
    return data

def run_cmd(cmd):
    '''cmd must be a list of arguments'''
    try:
        with open(os.devnull) as devnull:
            # is this the right way to block stdin?
            data = subprocess.check_output(cmd, stderr=subprocess.STDOUT, stdin=devnull)
        status = 0
    except subprocess.CalledProcessError as ex:
        data = ex.output
        status = ex.returncode
    data = data.decode('utf8')
    if status != 0:
        print('FAILED with status {}'.format(status))
        print('output was:')
        print(data)
        sys.exit(1)
    return data
